File: Data Structures & Algorithms/validate-parentheses/submission-1.java
class Solution {
    public boolean isValid(String s) {
        Stack<Character> stack = new Stack<>();
        HashMap<Character, Character> q = new HashMap<>();
        q.put('[', ']');
        q.put('(', ')');
        q.put('{', '}');

        for (int i = 0; i < s.length(); i++) {
            var character = s.charAt(i);
            if (q.containsKey(character))
                stack.push(character);
            else {
                if (stack.size() == 0)
                    return false;
                var closing = stack.pop();
                if (!q.get(closing).equals(character))
                    return false;
            }
        }
        return stack.isEmpty();
    }
}

File: Data Structures & Algorithms/generate-parentheses/submission-1.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        res = []
        stack = []

        def dfs(opening, closing):
            if closing > opening or opening > n:
                return  
            if opening == closing and opening == n:
                res.append("".join(stack))
            stack.append("(")
            dfs(opening + 1, closing)
            stack.pop()
            stack.append(")")
            dfs(opening, closing + 1)
            stack.pop()
        
        dfs(0, 0)
        return res